HPCC-27298 Add min and max skew columns to ECL watch file search results

jeclrsg opened the pull request:
Signed-off-by: Jeremy Clements <jeremy.clements@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-27298 Add min and max skew columns to ECL watch file search results

Signed-off-by: Jeremy Clements <jeremy.clements@lexisnexisrisk.com>

jeclrsg commented on esp/src/eclwatch/DFUQueryWidget.js:
```diff
@@ -673,7 +674,31 @@ define([
+                        }
+                    },
+                    Modified: { label: this.i18n.ModifiedUTCGMT, width: 162 },
+                    AtRestCost: {
```
I missed adding these fields to the dojo component in HPCC-27149, so just thought I'd add them now.

HPCCSmoketest commented:
Process of PR-15996, label: hpcc-27298-skew-columns-dfuquery-views starts now.
The reason of this test is: Forced to re-test.
Commit ID: 63d9a0e8ca81f32a1ea2b113e81c776f93840688
Estimated completion time is 0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-44.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-44.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 63d9a0e8ca81f32a1ea2b113e81c776f93840688
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.6.19-closedown0.el7.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  433 sec (00:07:13) |  109 sec (00:01:49) |   20 sec (00:00:20) |   16 sec (00:00:16) |    0 sec (00:00:00) |   14 sec (00:00:14) |  610 sec (00:10:10) |

GordonSmith commented on esp/src/eclwatch/DFUQueryWidget.js:
```diff
@@ -673,7 +674,31 @@ define([
+                    AtRestCost: {
+                        label: nlsHPCC.FileCostAtRest, width: 100,
+                        formatter: function (cost, row) {
+                            return Session.formatCost(cost || 0) + " (" + (dojoConfig.currencyCode || "$") + ")";
```
Minor: We should start using `??` instead of `||` in this scenario.

GordonSmith commented on esp/src/eclwatch/DFUQueryWidget.js:
```diff
@@ -673,7 +674,31 @@ define([
+                    AccessCost: {
+                        label: nlsHPCC.FileAccessCost, width: 100,
+                        formatter: function (cost, row) {
+                            return Session.formatCost(cost || 0) + " (" + (dojoConfig.currencyCode || "$") + ")";
```
Minor: We should start using `??` instead of `||` in this scenario.

GordonSmith replied:
`" (" + (dojoConfig.currencyCode || "$") + ")";` and related formatting helpers could be moved into Session file?

jeclrsg commented on esp/src/src/Session.ts:
```diff
@@ -45,7 +45,10 @@ getBuildInfo().then(info => {
-export const formatCost = d3Format(".2f");
+export function formatCost(value) {
+    const format = d3Format(".2f");
+    return `${format(value)} (${dojoConfig?.currencyCode || "$"})`;
```
Evidently ?? doesn't consider empty string something to instead return the right-hand side, so used the || instead

GordonSmith replied:
Correct...

HPCCSmoketest commented:
Process of PR-15996, label: hpcc-27298-skew-columns-dfuquery-views starts now.
The reason of this test is: Forced to re-test.
Commit ID: 58491e8c20ff19b22d491f9623b5fe14885fa129
Estimated completion time is 0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-77.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-77.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 58491e8c20ff19b22d491f9623b5fe14885fa129
Containerized:False
Number of Error(s): 5
src-react/components/LogicalFileSummary.tsx(161,107): error TS2339: Property AccessCost does not exist on type LogicalFile.
src-react/components/LogicalFileSummary.tsx(162,107): error TS2339: Property AtRestCost does not exist on type LogicalFile.
make[2]: *** [esp/src/CMakeFiles/eclwatch.dir/build.make:330: esp/src/build/dist/stub.eclwatch.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:12240: esp/src/CMakeFiles/eclwatch.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-04-13_17-41-54
time logs started
Prep time   :    23 sec (00:00:23)
Build time  :   167 sec (00:02:47)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   190 sec (00:03:10)
time logs end

[Build error:](https://gist.github.com/5eb59899acc84e5d07451be734d8040d)
Build log [Build log](https://gist.github.com/HPCCSmoketest/5eb59899acc84e5d07451be734d8040d#file-result-22-04-13-17-38-26-log)

GordonSmith commented on esp/src/src/Session.ts:
```diff
@@ -45,7 +45,10 @@ getBuildInfo().then(info => {
 
-export const formatCost = d3Format(".2f");
+export function formatCost(value) {
+    const format = d3Format(".2f");
```
The format function can be moved into module scope.

GordonSmith commented on esp/src/src/Session.ts:
```diff
@@ -45,7 +45,10 @@ getBuildInfo().then(info => {
 });
 
-export const formatCost = d3Format(".2f");
+export function formatCost(value) {
```
value should be typed as 
```ts
export function formatCost(value?:number|string):string {...}
```
And move the edge case handling into the function (empty string, undefined etc.).

jeclrsg replied:
I thought for falsy values it should return an empty string. And I added that logger line, but that might not be useful?

HPCCSmoketest commented:
Process of PR-15996, label: hpcc-27298-skew-columns-dfuquery-views starts now.
The reason of this test is: Forced to re-test.
Commit ID: 83fd0e713020fa858b572937b9e3dd2d56fee845
Estimated completion time is 0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-86.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-86.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 83fd0e713020fa858b572937b9e3dd2d56fee845
Containerized:False
Number of Error(s): 5
src-react/components/LogicalFileSummary.tsx(161,107): error TS2339: Property AccessCost does not exist on type LogicalFile.
src-react/components/LogicalFileSummary.tsx(162,107): error TS2339: Property AtRestCost does not exist on type LogicalFile.
make[2]: *** [esp/src/CMakeFiles/eclwatch.dir/build.make:330: esp/src/build/dist/stub.eclwatch.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:12240: esp/src/CMakeFiles/eclwatch.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-04-14_15-57-36
time logs started
Prep time   :    17 sec (00:00:17)
Build time  :   159 sec (00:02:39)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   176 sec (00:02:56)
time logs end

[Build error:](https://gist.github.com/b56d8265c3fcdaca68929a34dcc0adec)
Build log [Build log](https://gist.github.com/HPCCSmoketest/b56d8265c3fcdaca68929a34dcc0adec#file-result-22-04-14-15-54-24-log)

GordonSmith commented on esp/src/eclwatch/DFUQueryWidget.js:
```diff
@@ -673,7 +674,31 @@ define([
+                    AccessCost: {
+                        label: nlsHPCC.FileAccessCost, width: 100,
+                        formatter: function (cost, row) {
+                            return Session.formatCost(cost || 0);
```
No need for the `|| 0` now?

GordonSmith commented on esp/src/eclwatch/DFUQueryWidget.js:
```diff
@@ -673,7 +674,31 @@ define([
+                    AtRestCost: {
+                        label: nlsHPCC.FileCostAtRest, width: 100,
+                        formatter: function (cost, row) {
+                            return Session.formatCost(cost || 0);
```
No need for the `|| 0` now?

GordonSmith commented on esp/src/eclwatch/LFDetailsWidget.js:
```diff
@@ -461,7 +461,7 @@ define([
                 this.updateInput("RecordSize", oldValue, this.i18n.NoPublishedSize);
             } else if (name === "Cost") {
-                this.updateInput("FormattedCost", oldValue, Session.formatCost(newValue || 0) + " (" + dojoConfig.currencyCode + ")");
+                this.updateInput("FormattedCost", oldValue, Session.formatCost(newValue || 0));
```
No need for the `|| 0` now?

GordonSmith commented on esp/src/eclwatch/WUDetailsWidget.js:
```diff
@@ -609,9 +609,9 @@ define([
                 domElem.set("value", newValue.Item.join("\n"));
             } else if (name === "ExecuteCost") {
-                this.updateInput("FormattedExecuteCost", oldValue, Session.formatCost(newValue || 0) + " (" + dojoConfig.currencyCode + ")");
+                this.updateInput("FormattedExecuteCost", oldValue, Session.formatCost(newValue || 0));
```
No need for the `|| 0` now?

GordonSmith commented on esp/src/eclwatch/WUDetailsWidget.js:
```diff
@@ -609,9 +609,9 @@ define([
+                this.updateInput("FormattedExecuteCost", oldValue, Session.formatCost(newValue || 0));
             } else if (name === "FileAccessCost") {
-                this.updateInput("FormattedFileAccessCost", oldValue, Session.formatCost(newValue || 0) + " (" + dojoConfig.currencyCode + ")");
+                this.updateInput("FormattedFileAccessCost", oldValue, Session.formatCost(newValue || 0));
```
No need for the `|| 0` now?

GordonSmith commented on esp/src/src-react/components/Files.tsx:
```diff
@@ -162,17 +160,23 @@ export const Files: React.FunctionComponent<FilesProps> = ({
                 label: nlsHPCC.FileCostAtRest, width: 100,
                 formatter: function (cost, row) {
-                    return `${formatCost(cost ?? 0)} (${currencyCode || "$"})`;
+                    return `${formatCost(cost ?? 0)}`;
```
No need for the `?? 0` now?

GordonSmith commented on esp/src/src-react/components/Files.tsx:
```diff
@@ -162,17 +160,23 @@ export const Files: React.FunctionComponent<FilesProps> = ({
                 label: nlsHPCC.FileAccessCost, width: 100,
                 formatter: function (cost, row) {
-                    return `${formatCost(cost ?? 0)} (${currencyCode || "$"})`;
+                    return `${formatCost(cost ?? 0)}`;
```
No need for the `?? 0` now?

GordonSmith commented on esp/src/src-react/components/LogicalFileSummary.tsx:
```diff
@@ -160,7 +158,8 @@ export const LogicalFileSummary: React.FunctionComponent<LogicalFileSummaryProps
                 "JobName": { label: nlsHPCC.JobName, type: "string", value: file?.JobName, readonly: true },
-                "Cost": { label: nlsHPCC.Cost, type: "string", value: `${formatCost(file?.Cost ?? 0)} (${currencyCode})`, readonly: true },
+                "AccessCost": { label: nlsHPCC.FileAccessCost, type: "string", value: `${formatCost(file?.AccessCost ?? 0)}`, readonly: true },
+                "AtRestCost": { label: nlsHPCC.FileCostAtRest, type: "string", value: `${formatCost(file?.AtRestCost ?? 0)}`, readonly: true },
```
No need for the `?? 0` now?

GordonSmith commented on esp/src/src-react/components/Workunits.tsx:
```diff
@@ -140,14 +138,13 @@ export const Workunits: React.FunctionComponent<WorkunitsProps> = ({
                 formatter: function (cost, row) {
-                    return `${formatCost(cost ?? 0)
-                        } (${currencyCode || "$"})`;
+                    return `${formatCost(cost ?? 0)}`;
```
No need for the `?? 0` now?

GordonSmith commented on esp/src/src-react/components/Workunits.tsx:
```diff
@@ -140,14 +138,13 @@ export const Workunits: React.FunctionComponent<WorkunitsProps> = ({
                 label: nlsHPCC.ExecuteCost, width: 100,
                 formatter: function (cost, row) {
-                    return `${formatCost(cost ?? 0)} (${currencyCode || "$"})`;
+                    return `${formatCost(cost ?? 0)}`;
```
No need for the `?? 0` now?

GordonSmith commented on esp/src/src-react/components/WorkunitSummary.tsx:
```diff
@@ -179,8 +177,8 @@ export const WorkunitSummary: React.FunctionComponent<WorkunitSummaryProps> = ({
-                                "executeCost": { label: nlsHPCC.ExecuteCost, type: "string", value: `${formatCost(workunit?.ExecuteCost ?? 0)} (${currencyCode})`, readonly: true },
-                                "fileAccessCost": { label: nlsHPCC.FileAccessCost, type: "string", value: `${formatCost(workunit?.FileAccessCost ?? 0)} (${currencyCode})`, readonly: true },
+                                "executeCost": { label: nlsHPCC.ExecuteCost, type: "string", value: `${formatCost(workunit?.ExecuteCost ?? 0)}`, readonly: true },
+                                "fileAccessCost": { label: nlsHPCC.FileAccessCost, type: "string", value: `${formatCost(workunit?.FileAccessCost ?? 0)}`, readonly: true },
```
No need for the `?? 0` now?

HPCCSmoketest commented:
Process of PR-15996, label: hpcc-27298-skew-columns-dfuquery-views starts now.
The reason of this test is: Forced to re-test.
Commit ID: e08b46e2d9a37a4f7d5eafb5b66309c252cce0bc
Estimated completion time is 0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-133.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2

jeclrsg commented:
@GordonSmith I made the last changes you requested. I also squashed and rebased to the newest 8.6.x. That seems to make that "Compare" button a little weird until you change the left hand side of the comparison to the candidate.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-133.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: e08b46e2d9a37a4f7d5eafb5b66309c252cce0bc
Containerized:False
Number of Error(s): 5
src-react/components/LogicalFileSummary.tsx(161,107): error TS2339: Property AccessCost does not exist on type LogicalFile.
src-react/components/LogicalFileSummary.tsx(162,107): error TS2339: Property AtRestCost does not exist on type LogicalFile.
make[2]: *** [esp/src/CMakeFiles/eclwatch.dir/build.make:330: esp/src/build/dist/stub.eclwatch.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:12240: esp/src/CMakeFiles/eclwatch.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-04-14_21-50-51
time logs started
Prep time   :    18 sec (00:00:18)
Build time  :   169 sec (00:02:49)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   187 sec (00:03:07)
time logs end

[Build error:](https://gist.github.com/3ef1400c9988e104a724fc5ea9014d7e)
Build log [Build log](https://gist.github.com/HPCCSmoketest/3ef1400c9988e104a724fc5ea9014d7e#file-result-22-04-14-21-47-27-log)

HPCCSmoketest commented:
Process of PR-15996, label: hpcc-27298-skew-columns-dfuquery-views starts now.
The reason of this test is: Forced to re-test.
Commit ID: ebf42f32328896963f130eb82ba99c75419e7a24
Estimated completion time is 0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-22.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-22.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: ebf42f32328896963f130eb82ba99c75419e7a24
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.6.21-closedown0.el7.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  407 sec (00:06:47) |  100 sec (00:01:40) |   22 sec (00:00:22) |   16 sec (00:00:16) |    0 sec (00:00:00) |   14 sec (00:00:14) |  575 sec (00:09:35) |
